test/helper/sending.bash

```diff
@@ -45,30 +35,111 @@ function _send_email() {
       ( '--server' ) SERVER=${2:?--server given but no argument} ; shift 2 ;;
       ( '--port' )   PORT=${2:?--port given but no argument}     ; shift 2 ;;
       ( '--data' )
-        local TEMPLATE_FILE="/tmp/docker-mailserver-test/emails/${2:?--data given but no argument provided}.txt"
-        FINAL_SWAKS_OPTIONS+=('--data')
-        FINAL_SWAKS_OPTIONS+=('-')
-        FINAL_SWAKS_OPTIONS+=('<')
-        FINAL_SWAKS_OPTIONS+=("${TEMPLATE_FILE}")
+        ADDITIONAL_SWAKS_OPTIONS+=('--data')
+        local FILE_PATH="/tmp/docker-mailserver-test/emails/${2:?--data given but no argument provided}"
+        if _exec_in_container_bash "[[ -e ${FILE_PATH} ]]"; then
+          ADDITIONAL_SWAKS_OPTIONS+=("@${FILE_PATH}")
+        else
+          ADDITIONAL_SWAKS_OPTIONS+=("'${2}'")
+        fi
         shift 2
+        DATA_WAS_SUPPLIED=1
         ;;
-      ( * ) ADDITIONAL_SWAKS_OPTIONS+=("${1}") ; shift 1 ;;
+      ( * ) ADDITIONAL_SWAKS_OPTIONS+=("'${1}'") ; shift 1 ;;
     esac
   done
 
-  _run_in_container_bash "swaks --server ${SERVER} --port ${PORT} --ehlo ${EHLO} --from ${FROM} --to ${TO} ${ADDITIONAL_SWAKS_OPTIONS[*]} ${FINAL_SWAKS_OPTIONS[*]}"
+  if [[ ${DATA_WAS_SUPPLIED} -eq 0 ]]; then
+    # Fallback template without implicit `Message-Id` + `X-Mailer` headers:
+    # NOTE: It is better to let Postfix generate the `Message-Id` header and append that
+    #       as it contains the queue ID for tracking logs and is returned to in swaks output.
+    ADDITIONAL_SWAKS_OPTIONS+=('--data')
+    ADDITIONAL_SWAKS_OPTIONS+=("'Date: %DATE%\nTo: %TO_ADDRESS%\nFrom: %FROM_ADDRESS%\nSubject: test %DATE%\n%NEW_HEADERS%\n%BODY%\n'")
+  fi
+
+  echo "swaks --server '${SERVER}' --port '${PORT}' --ehlo '${EHLO}' --from '${FROM}' --to '${TO}' ${ADDITIONAL_SWAKS_OPTIONS[*]}"
+}
+
+# Sends a mail from the container named by the environment variable `CONTAINER_NAME`
+# to the same or another container.
+#
+# To send a custom email, create a file at `test/files/<TEST FILE>`,
+# and provide `<TEST FILE>` as an argument to this function.
+#
+# Parameters include all options that one can supply to `swaks` itself.
+# The `--data` parameter expects either, a relative path from `emails/` or
+# "inline" data, e.g., `Date: 1 Jan 2024\nSubject: This is a test`.
+#
+# ## Output
+#
+# This functions prints the output of the transaction that `swaks` prints.
+#
+# ## Attention
+#
+# This function assumes `CONTAINER_NAME` to be properly set (to the container
+# name the command should be executed in)!
+#
+# This function will send the email in an "asynchronous" fashion,
+# it will return without waiting for the Postfix mail queue to be emptied.
+#
+# This functions performs **no** implicit `assert_success` to check whether
+# the e-mail transaction was successful. If this is not desirable, use
+# `_send_email`. If you anticipate the sending to succeed, use `_send_email`
+# instead.
```

Comment:
I actually have a 34" 21:9 display - line length should not be an issue right? 😆 As a matter of fact though, I am now used to having 3 coloums side by side, and even on such a monitor, it can get crowded on the display..